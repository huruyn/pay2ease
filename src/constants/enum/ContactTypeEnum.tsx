// eslint-disable-next-line import/prefer-default-export
export enum ContactTypeEnum {
  PERSONAL = 'personal',
  BUSINESS = 'business',
}
