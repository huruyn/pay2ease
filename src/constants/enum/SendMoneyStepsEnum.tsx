// eslint-disable-next-line import/prefer-default-export
export enum SendMoneyStepsEnum {
  TRANSACTION_START = 'transaction-start',
  TRANSACTION_DETAILS = 'transaction-details',
  TRANSACTION_SECURITY_QUESTIONS = 'transaction-security-questions',
  TRANSACTION_STATUS = 'transaction-status',
  TRANSACTION_COMPLETE = 'transaction-complete',
}
