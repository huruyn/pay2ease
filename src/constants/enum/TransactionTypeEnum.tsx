// eslint-disable-next-line import/prefer-default-export
export enum TransactionTypeEnum {
  SEND = 'send',
  REQUEST = 'request',
}
