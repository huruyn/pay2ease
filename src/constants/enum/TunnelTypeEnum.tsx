// eslint-disable-next-line import/prefer-default-export
export enum TunnelTypeEnum {
  ZIPPY_CASH = 'Zippy.Cash',
  INTERAC_E_TRANSFER = 'InteracÂ® e-Transfer',
}
