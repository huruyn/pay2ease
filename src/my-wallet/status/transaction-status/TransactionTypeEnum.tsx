// eslint-disable-next-line import/prefer-default-export
export enum TransactionTypeEnum {
  SENT = 'sent',
  RECEIVED = 'received',
  REQUESTED = 'requested',
}
