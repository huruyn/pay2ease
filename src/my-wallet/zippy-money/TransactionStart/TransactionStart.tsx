import { Button, Form } from 'react-bootstrap';
import { useNavigate } from 'react-router-dom';
import PageContainer from '../../../common/PageContainer';
import { TransactionInterface } from '../../../constants/interface/TransactionInterface';
import { TransactionTypeEnum } from '../../../constants/enum/TransactionTypeEnum';
import { TunnelTypeEnum } from '../../../constants/enum/TunnelTypeEnum';
import ContactInput from '../../../common/ContactInput/ContactInput';
import AmountInput from '../../../common/AmountInput';

export default function TransactionStart({
  contactList,
  setContactList,
  // accountsList,
  transactionType,
  handleTriggerTransaction,
  isProcessing,
  // errorMessage,
  mainInfo,
  setMainInfo,
  // selectedAccount,
  // setSelectedAccount,
  selectedContact,
  setSelectedContact,
  tunnelType,
  setTunnelType,
}: TransactionInterface) {
  const navigate = useNavigate();

  const validate = () => transactionType && selectedContact.id && mainInfo.amount && tunnelType;

  return (
    <>
      <PageContainer
        title="Personal Account"
        subTitle="Made Fun With Zippy!"
        backdropImage="backdrop-image-2"
      >
        <div className="zippy-btn-group btn-group" role="group" aria-label="Zippy Cash">
          <Button
            className={transactionType === TransactionTypeEnum.SEND ? 'active' : ''}
            onClick={() => navigate(`/my-wallet/zippy-money/${TransactionTypeEnum.SEND}/transaction-start`)}
            disabled={transactionType === TransactionTypeEnum.SEND}
          >
            Send Money
          </Button>
          <Button
            className={transactionType === TransactionTypeEnum.REQUEST ? 'active' : ''}
            onClick={() => navigate(`/my-wallet/zippy-money/${TransactionTypeEnum.REQUEST}/transaction-start`)}
            disabled={transactionType === TransactionTypeEnum.REQUEST}
          >
            Request Money
          </Button>
        </div>
        <ContactInput
          contactList={contactList}
          setContactList={setContactList}
          selectedContact={selectedContact}
          setSelectedContact={setSelectedContact}
        />
        <AmountInput
          amount={mainInfo.amount}
          setAmount={(value) => setMainInfo({ ...mainInfo, amount: Number(value) })}
        />
        <div className="select-method">
          <Form.Check
            id="zippy-cash"
            type="radio"
            className="actions"
            label="Zippy.Cash"
            name="actions2"
            onChange={() => setTunnelType(TunnelTypeEnum.ZIPPY_CASH)}
            checked={tunnelType === TunnelTypeEnum.ZIPPY_CASH}
            disabled
          />
          <Form.Check
            id="interac-e-transfer"
            type="radio"
            className="actions"
            label="InteracÂ® e-Transfer"
            name="actions2"
            onChange={() => setTunnelType(TunnelTypeEnum.INTERAC_E_TRANSFER)}
            checked={tunnelType === TunnelTypeEnum.INTERAC_E_TRANSFER}
          />
        </div>
        <div className="action">
          <Button
            className="zippy-btn"
            disabled={isProcessing || !validate()}
            onClick={handleTriggerTransaction}
          >
            {isProcessing && <div className="loading spinner-border" role="status" />}
            Zipp It
          </Button>
        </div>
      </PageContainer>
    </>
  );
}
