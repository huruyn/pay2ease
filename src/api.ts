/* eslint-disable no-console */

import { AccountInfo, InteractionRequiredAuthError, IPublicClientApplication } from '@azure/msal-browser';
import { TransactionStatusEnum } from './interac-etransfer/status/transaction-status/TransactionStatusEnum';

export enum TransferType {
  SEND = 'send',
  RECEIVE = 'receive',
  REQUEST = 'request',
}

export type Account = {
  id: number;
  name: string;
  balance: number;
};

export type ContactBase = {
  firstName: string;
  lastName: string;
  email: string;
  nickname?: string;
  phone?: string;
};

export type Contact = ContactBase & {
  id: number;
};

export type User = {
  email: string;
  firstName: string;
  lastName: string;
};

export type Transaction = {
  status: TransactionStatusEnum;
  id: number;
  contact: Contact;
  amount: number;
  date: string;
  expireDate: string;
  securityQuestion?: string;
  securityAnswer?: string;
};

export type InteracEtransferTransaction = {
  id?: number,
  contactId: number,
  amount: number,
  type: string,
  securityQuestion?: string,
  securityAnswer?: string,
};

export type ZippyCashTransaction = {
  id?: number,
  contactId: number,
  amount: number,
  isCredit: boolean,
  createdDate?: string
};

export type VersionResponse = {
  version: string;
};

export type ErrorDetail = {
  type: string,
  title: string,
  status: number,
  traceId: string,
};

export async function getToken(instance: IPublicClientApplication, account: AccountInfo)
  : Promise<string | null> {
  const accessTokenRequest = {
    scopes: ['https://zippycashdev.onmicrosoft.com/b0b070aa-4e90-4015-af46-59d0ceed5ecc/access_as_user'],
    account,
  };

  let accessToken;
  try {
    const accessTokenResponse = await instance.acquireTokenSilent(accessTokenRequest);
    accessToken = accessTokenResponse.accessToken;
    return accessToken;
  } catch (error) {
    if (error instanceof InteractionRequiredAuthError) {
      await instance.acquireTokenRedirect(accessTokenRequest);
    }
    console.log(error);
  }
  return null;
}

export default class Api {
  constructor(
    private readonly instance: IPublicClientApplication,
    private readonly account: AccountInfo,
  ) { }

  public listAccounts() {
    return this.fetch<Account[]>('get', 'Accounts');
  }

  public listContacts() {
    return this.fetch<Contact[]>('get', 'Contacts');
  }

  public postContact(data: ContactBase) {
    return this.fetch<Contact>('post', 'Contacts', data);
  }

  public putUser() {
    return this.fetch<User>('put', 'User');
  }

  public getInteracEtransferTransactions(type: TransferType) {
    return this.fetch<Transaction[]>('get', `InteracEtransfer/Transactions?type=${type}`);
  }

  public getInteracEtransferTransaction(transactionId: number) {
    return this.fetch<Transaction>('get', `InteracEtransfer/Transactions/${transactionId}`);
  }

  public getInteracEtransferCancelTransaction(transactionId: number) {
    return this.fetch<Transaction>('post', `InteracEtransfer/Transactions/${transactionId}/Cancel`);
  }

  public getInteracEtransferSendReminder(transactionId: number) {
    return this.fetch<Transaction>('post', `InteracEtransfer/Transactions/${transactionId}/Remind`);
  }

  public postInteracEtransferTransaction(data: InteracEtransferTransaction) {
    return this.fetch<InteracEtransferTransaction>('post', 'InteracEtransfer/Transactions', data);
  }

  public postDirectDepositStatus(email: string) {
    return this.fetch<InteracEtransferTransaction>('post', 'InteracEtransfer/DirectDepositStatus', { email });
  }

  public getZippyCashTransfer(transactionId: number) {
    return this.fetch<Transaction>('get', `Transfers/${transactionId}`);
  }

  public postZippyCashTransfer(data: ZippyCashTransaction) {
    return this.fetch<ZippyCashTransaction>('post', 'Transfers', data);
  }

  public getVersion() {
    return this.fetch<VersionResponse>('get', 'Version');
  }

  private async fetch<TResponse>(method: string, path: string, body?: any) {
    const apiUrl = (<any>window).API_URL;
    if (!apiUrl) {
      throw Error('window.API_URL is undefined');
    }

    const request: RequestInit = {
      method,
      body: body ? JSON.stringify(body) : null,
    };

    if (path !== 'Version') {
      const accessToken = await getToken(this.instance, this.account);
      request.headers = new Headers({
        Authorization: `Bearer ${accessToken}`,
        'Content-Type': 'application/json',
      });
    }

    const response = await fetch(`${apiUrl}/${path}`, request);
    // TODO: proper error output here
    if (!response.ok) {
      const error = await response.json() as ErrorDetail;
      throw Error(error.title);
    }
    return (response.status !== 204 ? await response.json() : {}) as TResponse;
  }
}
