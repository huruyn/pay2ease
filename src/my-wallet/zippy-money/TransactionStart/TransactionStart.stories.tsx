import { ComponentStory, ComponentMeta } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';
import TransactionStart from './TransactionStart';
import MAIN_INFO from '../../../stories/MainInfo';
import { SendMoneyStepsEnum } from '../../../constants/enum/SendMoneyStepsEnum';
import { TransactionTypeEnum } from '../../../constants/enum/TransactionTypeEnum';
import { Account, Contact } from '../../../api';

export default {
  title: 'my-wallet/zippy-money/TransactionStart',
  component: TransactionStart,
} as ComponentMeta<typeof TransactionStart>;

const Template: ComponentStory<typeof TransactionStart> = () => (
  <BrowserRouter>
    <TransactionStart
      accountsList={[]}
      contactList={[]}
      setContactList={jest.fn()}
      errorMessage=""
      handleTriggerTransaction={undefined}
      isProcessing={false}
      mainInfo={MAIN_INFO}
      selectedAccount={{} as Account}
      selectedContact={{} as Contact}
      setMainInfo={jest.fn()}
      setSelectedAccount={jest.fn()}
      setSelectedContact={jest.fn()}
      step={SendMoneyStepsEnum.TRANSACTION_START}
      transactionType={TransactionTypeEnum.SEND}
      setTunnelType={jest.fn()}
    />
  </BrowserRouter>
);

export const Default = Template.bind({});
Default.args = {};
Default.parameters = {};

export const APIError = Template.bind({});
APIError.args = {};
APIError.parameters = {};
