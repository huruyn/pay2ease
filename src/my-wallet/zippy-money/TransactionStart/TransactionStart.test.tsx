import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import TransactionStart from './TransactionStart';
import MAIN_INFO from '../../../stories/MainInfo';
import { SendMoneyStepsEnum } from '../../../constants/enum/SendMoneyStepsEnum';
import { TransactionTypeEnum } from '../../../constants/enum/TransactionTypeEnum';
import { Account, Contact } from '../../../api';

const ReactTestRenderer = require('react-test-renderer');

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...(jest.requireActual('react-router-dom') as any),
  useParams: () => ({
    stepId: 'details',
  }),
  useNavigate: () => mockedUsedNavigate,
}));

describe('TransactionStart Component', () => {
  it('matches the snapshot', () => {
    const tree = ReactTestRenderer.create(
      <BrowserRouter>
        <TransactionStart
          accountsList={[]}
          contactList={[]}
          setContactList={jest.fn()}
          errorMessage=""
          handleTriggerTransaction={undefined}
          isProcessing={false}
          mainInfo={MAIN_INFO}
          selectedAccount={{} as Account}
          selectedContact={{} as Contact}
          setMainInfo={jest.fn()}
          setSelectedAccount={jest.fn()}
          setSelectedContact={jest.fn()}
          step={SendMoneyStepsEnum.TRANSACTION_START}
          transactionType={TransactionTypeEnum.SEND}
          setTunnelType={jest.fn()}
        />
      </BrowserRouter>,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
