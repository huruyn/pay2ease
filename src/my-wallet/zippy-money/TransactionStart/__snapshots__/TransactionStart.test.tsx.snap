// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionStart Component matches the snapshot 1`] = `
<div
  className=""
>
  <div
    className="zippy-cash-container-wrap"
  >
    <div
      className="backdrop backdrop-image-2"
    />
    <div
      className="zippy-cash-container"
    >
      <div
        className="zippy-cash-container-header"
      >
        <h1
          className="title"
        >
          Personal Account
        </h1>
        <h2
          className="catch"
        >
          Made Fun With Zippy!
        </h2>
      </div>
      <div
        className="zippy-cash-container-body"
      >
        <div
          aria-label="Zippy Cash"
          className="zippy-btn-group btn-group"
          role="group"
        >
          <button
            className="active btn btn-primary"
            disabled={true}
            onClick={[Function]}
            type="button"
          >
            Send Money
          </button>
          <button
            className="btn btn-primary"
            disabled={false}
            onClick={[Function]}
            type="button"
          >
            Request Money
          </button>
        </div>
        <div
          className="contact-selector"
        >
          <div
            className="contact-identifier"
          >
            <input
              className="form-control"
              onChange={[Function]}
              placeholder="Enter @ZippyUsername or Email"
              type="text"
            />
            <button
              className="btn-select-contact btn-transparent btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <i
                className="zippy-cash-icon zc-add"
              />
            </button>
          </div>
          <div
            className="contact-list-wrap"
          >
            <div
              className="contact-container hidden"
            >
              <div
                aria-label="Zippy Cash"
                className="zippy-btn-group btn-group"
                role="group"
              >
                <button
                  className="active btn btn-primary"
                  disabled={true}
                  onClick={[Function]}
                  type="button"
                >
                  Personal
                </button>
                <button
                  className="btn btn-primary"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  Business
                </button>
              </div>
              <div
                className="contact-search"
              >
                <i
                  className="zippy-cash-icon zc-search"
                />
                <input
                  className="form-control"
                  onChange={[Function]}
                  placeholder="Search..."
                  value=""
                />
              </div>
              <div
                className="contact-list-container"
              >
                <div>
                  Loading...
                </div>
              </div>
              <div
                className="action"
              >
                <button
                  className="zippy-btn btn btn-primary"
                  disabled={true}
                  onClick={[Function]}
                  type="button"
                >
                  <div
                    className="loading spinner-border"
                    role="status"
                  />
                  Add Contact
                </button>
              </div>
            </div>
            <div
              className="contact-container hidden"
            >
              <div
                className="container-header"
              >
                <h2
                  className="title"
                >
                  Add  Contact
                </h2>
              </div>
              <div
                className="container-body"
              >
                <input
                  className="common-input form-control"
                  name="firstname"
                  onChange={[Function]}
                  placeholder="Firstname"
                  type="text"
                  value=""
                />
                <input
                  className="common-input form-control"
                  name="lastname"
                  onChange={[Function]}
                  placeholder="Lastname"
                  type="text"
                  value=""
                />
                <input
                  className="common-input form-control"
                  name="email"
                  onChange={[Function]}
                  placeholder="Email"
                  type="text"
                  value=""
                />
                <input
                  className="common-input form-control"
                  name="nickname"
                  onChange={[Function]}
                  placeholder="Nickname (optional)"
                  type="text"
                  value=""
                />
              </div>
              <div
                className="action"
              >
                <button
                  className="zippy-btn btn btn-primary"
                  disabled={true}
                  onClick={[Function]}
                  type="button"
                >
                  Add Contact
                </button>
              </div>
            </div>
          </div>
        </div>
        <input
          className="amount form-control"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="$0.00"
          type="string"
          value=""
        />
        <div
          className="select-method"
        >
          <div
            className="actions form-check"
          >
            <input
              checked={false}
              className="form-check-input"
              disabled={true}
              id="zippy-cash"
              name="actions2"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="form-check-label"
              htmlFor="zippy-cash"
              title=""
            >
              Zippy.Cash
            </label>
          </div>
          <div
            className="actions form-check"
          >
            <input
              checked={false}
              className="form-check-input"
              disabled={false}
              id="interac-e-transfer"
              name="actions2"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="form-check-label"
              htmlFor="interac-e-transfer"
              title=""
            >
              InteracÂ® e-Transfer
            </label>
          </div>
        </div>
        <div
          className="action"
        >
          <button
            className="zippy-btn btn btn-primary"
            disabled={true}
            type="button"
          >
            Zipp It
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
